repo/blob/sharded/sharded.go

```diff
@@ -178,8 +178,11 @@ func (s *Storage) GetMetadata(ctx context.Context, blobID blob.ID) (blob.Metadat
 
 // PutBlob implements blob.Storage.
 func (s *Storage) PutBlob(ctx context.Context, blobID blob.ID, data blob.Bytes, opts blob.PutOptions) error {
-	if opts.HasRetentionOptions() {
+	switch {
+	case opts.HasRetentionOptions():
 		return errors.New("setting blob-retention is not supported")
+	case opts.DoNotRecreate:
+		return errors.New("setting blob do-not-recreate is not supported")
```

Comment:
Wouldn't this be a function of the actual storage backend? It is true that currently the sharded provider is only used with storage backends that do not support DoNotRecreate, but still, it seems that it should be relayed to the backend.

Reply:
Good catch!

Yes, my bad. I should have taken `opts.HasRetentionOpts()` down to the three implementations of the `shared.Impl.PutBlobInPath` interface.

@adowair Can you move both the checks to the storage backend implementations of the shared interface ?

Reply:
It'd may be a good idea to break down and restructure the PR in self-contained changes (commits) that can be individually reviewed. For example, the change about passing down the options to the storage backends can be one of those steps. Merely a suggestion.

Reply:
Yes, that's the right way to do things. Rebasing upstream.